crates/lair_keystore_api/src/sodium_secretstream/framed.rs

```diff
@@ -0,0 +1,101 @@
+use super::*;
+
+/// wrap our streams with framing so we know message boundaries.
+pub(crate) fn priv_framed(
+    send: PrivRawSend,
+    recv: PrivRawRecv,
+) -> (PrivFramedSend, PrivFramedRecv) {
+    let send = PrivFramedSend(send);
+    let recv = PrivFramedRecv::new(recv);
+    (send, recv)
+}
+
+/// A framed sender.
+pub(crate) struct PrivFramedSend(PrivRawSend);
+
+impl PrivFramedSend {
+    /// Send framed data.
+    pub(crate) fn send(
+        &mut self,
+        d: Box<[u8]>,
+    ) -> impl Future<Output = LairResult<()>> + '_ + Send {
+        async move {
+            if d.len() > MAX_FRAME {
+                return Err(OneErr::with_message(
+                    "FrameOverflow",
+                    format!("{} > {}", d.len(), MAX_FRAME),
+                ));
+            }
+
+            let ltag = (d.len() as u16).to_le_bytes();
+
+            // something more efficient than just writing both buffers?
+
+            self.0.write_all(&ltag).await?;
+            self.0.write_all(&d).await?;
+            Ok(())
+        }
+    }
+
+    /// Forwards shutdown to the underlying raw sender.
+    pub(crate) fn shutdown(
+        &mut self,
+    ) -> impl Future<Output = LairResult<()>> + '_ + Send {
+        async move { self.0.shutdown().await.map_err(OneErr::new) }
+    }
+}
+
+/// A framed receiver.
+pub(crate) struct PrivFramedRecv(BoxStream<'static, LairResult<Box<[u8]>>>);
+
+impl PrivFramedRecv {
+    /// Initialize the framed receiver.
+    pub fn new(recv: PrivRawRecv) -> Self {
+        let recv = futures::stream::try_unfold(recv, |mut recv| async move {
+            // something more efficient than doing this in 2 steps?
+
+            // first, receive the u16_le frame length
+            let mut ltag = [0; 2];
+            recv.read_exact(&mut ltag).await?;
+            let ltag = u16::from_le_bytes(ltag) as usize;
+
+            // check if the length tag is out of bounds
+            if ltag > MAX_FRAME {
+                return Err(OneErr::with_message(
+                    "FrameOverflow",
+                    format!("{} > {}", ltag, MAX_FRAME),
+                ));
+            }
+
+            // initialize our receive buffer
+            let mut msg = Vec::with_capacity(ltag);
+
+            // we could try to figure out the async uninit thing
+            // (https://docs.rs/tokio/1.11.0/tokio/io/struct.ReadBuf.html#method.uninit)
+            // but this is actually safe given the reader doesn't inspect
+            // the uninitialized data, and in the case of an error
+            // this potentially partially filled buffer is not returned.
+            #[allow(unsafe_code)]
+            unsafe {
+                msg.set_len(ltag);
```

Comment:
What happens if this is partially filled and then dropped due to the error. Is the data cleaned up or leaked?

Reply:
The length is owned by the vec still, so it'll be cleaned up. The only unsafe part of set_len is that it can expose uninitialized bytes, which this function can't.